src/runtime_src/core/pcie/driver/linux/xocl/subdev/hwmon_sdm.c

```diff
@@ -0,0 +1,814 @@
+			}
+			if(srec->threshold_support_byte & THRESHOLD_LOWER_FATAL_MASK) {
+				xocl_dbg(&sdm->pdev->dev, "Lower Fatal Limit: 0x%x\r\n", *(srec->lower_fatal_limit));
+				xocl_dbg(&sdm->pdev->dev, "Lower Fatal Adjusted Limit : %f\r\n", (*(srec->lower_fatal_limit)) *
+					   pow(10, srec->unit_modifier_byte));
+			}
+		}
+		if(srec->base_unit_type_length != SDR_NULL_BYTE) {
+			uint8_t bu_type   = (srec->base_unit_type_length >> SDR_TYPE_POS) & SDR_TYPE_MASK;
+			if(bu_type == TYPECODE_ASCII)
+				xocl_dbg(&sdm->pdev->dev, "Base Unit Value : %s\r\n", srec->base_unit);
+			else if (bu_type == TYPECODE_BINARY)
+				xocl_dbg(&sdm->pdev->dev, "Sensor Base Unit : 0x%x\r\n", *(srec->base_unit));
+		}
+
+		uint8_t val_len = srec->value_type_length & SDR_LENGTH_MASK;
+		if(srec->unit_modifier_byte != 0x7F)
+		{
+			if(val_len == 1)
+				xocl_dbg(&sdm->pdev->dev, "Sensor Adjusted Value : %f\r\n", (*(uint8_t *)srec->value) *
+					   pow(10, srec->unit_modifier_byte));
+			else if(val_len == 2)
+				xocl_dbg(&sdm->pdev->dev, "Sensor Adjusted Value : %f\r\n", (*(uint16_t *)srec->value) *
+					   pow(10, srec->unit_modifier_byte));
+			else if(val_len == 4)
+				xocl_dbg(&sdm->pdev->dev, "Sensor Adjusted Value : %f\r\n", (*(uint32_t *)srec->value) *
+					   pow(10, srec->unit_modifier_byte));
+		}
+
+		if(srec->status == SENSOR_NOT_PRESENT)
+			xocl_dbg(&sdm->pdev->dev, "Sensor Status : Sensor Not Present\r\n");
+		else if(srec->status == SENSOR_PRESENT_AND_VALID)
+			xocl_dbg(&sdm->pdev->dev, "Sensor Status : Sensor Present and Valid\r\n");
+		else if(srec->status == DATA_NOT_AVAILABLE)
+			xocl_dbg(&sdm->pdev->dev, "Sensor Status : Sensor Data Not Available\r\n");
+		else if(srec->status == SENSOR_STATUS_NOT_AVAILABLE)
+			xocl_dbg(&sdm->pdev->dev, "Sensor Status : Sensor Data Not Available\r\n");
+
+		srec++;
+		sr--;
+
+		xocl_dbg(&sdm->pdev->dev, "\r\n");
+	}
+	xocl_dbg(&sdm->pdev->dev, "%s\r\n",  repo_object->eor.eor_marker);
+}
+
+static uint8_t sdr_get_id(uint8_t repo_type)
+{
+	int id = 0;
+
+	switch(repo_type) {
+		case SDR_TYPE_BDINFO: id = 0; break;
+		case SDR_TYPE_TEMP: id = 1; break;
+		case SDR_TYPE_VOLTAGE: id = 2; break;
+		case SDR_TYPE_CURRENT: id = 3; break;
+		case SDR_TYPE_POWER: id = 4; break;
+		case SDR_TYPE_QSFP: id = 5; break;
+		case SDR_TYPE_VPD_PCIE: id = 6; break;
+		case SDR_TYPE_IPMIFRU: id = 7; break;
+		case SDR_TYPE_CSDR_LOGDATA: id = 8; break;
+		case SDR_TYPE_VMC_LOGDATA: id = 9; break;
+		default: id = -1; break;
+	}
+	return id;
+}
+
+static struct sdr_response* parse_sdr_info(char *in_buf,
+                                           struct xocl_hwmon_sdm *sdm,
+                                           bool create_sysfs)
+{
+	int buf_index = SDR_NULL_BYTE;
+	uint8_t remaining_records = 0;
+	uint8_t srecords = 0;
+	uint8_t completion_code = in_buf[buf_index++];//don't change order
+	uint8_t repo_type = in_buf[buf_index];//don't change order
+	uint8_t repo_id = 0;
+	size_t sbyte = sizeof(uint8_t);
+
```

Comment:
Overall, the other code looks good to me, but this function needs more attention.
I understand that we use Sandeep's code as example, but it has to be in an acceptable quality. May we set up a zoom meeting and review this function together ASAP?

@chvamshi-xilinx @houlz0507 I think this function needs more work.

Reply:
After meeting with Raj, I think the fact is that this function is given by SC team as a standard parser for current sensor data format. The best way to move this project forward is that we accept current version that SC team is using and also investigate if we should make the sensor data data structure better in the future releases. This could be an ongoing process and needs 2 teams work together. Therefore, I'd recommend accept current version. Leave some comments/concerns for future improvement.